aeppl/scan.py

```diff
@@ -0,0 +1,327 @@
+
+    # We could clone `var`, but reusing it will make things easier down the
+    # line (e.g. avoid the need to remap cloned variables)
+    # new_input_var = var.clone()
+    new_outer_input_var = var.clone()
+    if new_outer_input_var.name:
+        new_outer_input_var.name = new_outer_input_var.name.lower()
+
+    var_slices = [new_outer_input_var[b:e] for b, e in slice_seqs]
+    n_steps = at.min([at.shape(n)[0] for n in var_slices])
+
+    if output_scan_args.n_steps is None or replacing:
+        output_scan_args.n_steps = n_steps
+
+    output_scan_args.outer_in_seqs += [v[:n_steps] for v in var_slices]
+
+    if not replacing or add_nit_sot:
+        output_scan_args.outer_in_nit_sot += [n_steps]
+
+    if inner_out_fn:
+        output_scan_args.inner_out_nit_sot += [
+            inner_out_fn(
+                input_scan_args, old_inner_out_var, inner_in_var, output_scan_args
+            )
+        ]
+
+    return output_scan_args, new_outer_input_var
+
+
+def get_random_outer_outputs(scan_args: ScanArgs) -> List[TensorVariable]:
+    """Get the `MeasurableVariable` outputs of a `Scan` (well, its `ScanArgs`)."""
+    rv_vars = []
+    for n, oo in enumerate(scan_args.outer_outputs):
+        oo_info = scan_args.find_among_fields(oo)
+        io_type = oo_info.name[(oo_info.name.index("_", 6) + 1) :]
+        inner_out_type = "inner_out_{}".format(io_type)
+        io_var = getattr(scan_args, inner_out_type)[oo_info.index]
+        if io_var.owner and isinstance(io_var.owner.op, MeasurableVariable):
+            rv_vars.append((n, oo, io_var))
+    return rv_vars
+
+
+def construct_scan(scan_args: ScanArgs) -> List[TensorVariable]:
+    scan_op = Scan(scan_args.inner_inputs, scan_args.inner_outputs, scan_args.info)
+    scan_out = scan_op(*scan_args.outer_inputs)
+
+    if not isinstance(scan_out, list):
+        scan_out = [scan_out]
+
+    return scan_out
+
+
+def create_inner_out_logp(
+    input_scan_args: ScanArgs,
+    old_inner_out_var: TensorVariable,
+    new_inner_in_var: TensorVariable,
+    output_scan_args: ScanArgs,
+) -> TensorVariable:
+    """Create a log-likelihood inner-output for a `Scan`."""
+    logp = _logprob(
+        old_inner_out_var.owner.op, new_inner_in_var, *old_inner_out_var.owner.inputs
+    )
+    if new_inner_in_var.name:
+        logp.name = "logp({})".format(new_inner_in_var.name)
+    return logp
+
+
+@local_optimizer([Scan])
+def check_scan_logprob(fgraph, node):
+    """Check that a `Scan` `Op` can be assigned a log-probability.
+
+    If it can, replace it with a `MeasurableScan` that works with `logprob`.
+    """
+
+    if isinstance(node.op, MeasurableScan) or not isinstance(node.op, Scan):
+        return
+
+    # TODO: Determine which outer-outputs are measurable.
+    # Use `curr_scanargs` to connect inner-outputs to outer-outputs.  If we
+    # can determine that an inner-output is a `MeasurableVariable`, then we
+    # can add the corresponding outer-output to `rv_outputs`.
```

Comment:
Without this in place, we're assuming that every `Scan` is a `MeasurableScan` (because we turn every `Scan` into one below), but that's not necessarily true.

We can apply the `Scan` transforms here and then convert the output into a `MeasurableScan`, or we can attempt to determine if a `Scan` can be transformed and do that later.  The former seems easier, though.

Reply:
Actually, it looks like we're letting `push_out_rvs_from_scan` run alongside this rewrite, and that should probably produce `MeasurableScan`s.